Boost: Add annotations support to performance history graph

haqadn opened the pull request:
## Proposed changes:
Changing how scores are calculated will confuse users. So, we need a way to communicate this change. This PR adds annotations support to performance history graph to add annotations for this and anything else.


https://github.com/Automattic/jetpack/assets/3737780/8b6a90ca-13d4-4269-85f0-c06f2ad65576



### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-2rd-p2#comment-1910

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
* Make sure performance history still loads.
* Try [injecting some annotations](https://github.com/Automattic/jetpack/blob/77648dc719780925765aa89d004a3d68ad5bc7a4/projects/plugins/boost/app/data-sync/Performance_History_Entry.php#L34) with performance history data in the following format:

```php
'annotations' => [
	[
		'timestamp' => 1704067200000,
		'text'      => 'Happy New Year!',
	]
]
```

haqadn committed:
Add annotations plugin

haqadn committed:
Add class prefix and fix z-index

haqadn committed:
Add annotation text element

haqadn committed:
Remove unused code

haqadn committed:
Add annotation pointer and set position

haqadn committed:
Add support for annotations

haqadn committed:
changelog

dilirity commented on projects/plugins/boost/app/assets/src/js/features/performance-history/lib/hooks.ts:
```diff
@@ -18,6 +18,14 @@ const periodsSchema = z.object( {
+					text: z.string(),
+				} )
+			)
+			.optional(),
```
The PHP schema for `annotations` is nullable and if PHP returns null, this will make JS sad. Though, I don't know when that might happen.

Either way, I noted it just to be safe, as we ran into this recently - https://github.com/Automattic/jetpack/pull/34932.

haqadn replied:
Thanks, I have made them not optional in the context of boost since we can actually control the back-end data. https://github.com/Automattic/jetpack/pull/34978/commits/e55a9505311af62d422aeae56549e9d78fbe3919

dilirity commented on projects/js-packages/components/components/boost-score-graph/tooltips-plugin.ts:
```diff
@@ -42,21 +43,15 @@ export function tooltipsPlugin( periods ) {
 			showTips();
 		} );
-
-		if ( u.cursor.left < 0 ) {
-			hideTips();
-		} else {
-			showTips();
-		}
```
Why did this get removed? Just curious.

haqadn replied:
The code didn't seem to have any useful effect and was confusing.

dilirity commented on projects/js-packages/components/components/boost-score-graph/annotations-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+import uPlot from 'uplot';
+import { Annotation } from '.';
+
+import './style-annotation.scss';
```
I was thinking whether it's a good idea to use global styles or local, but this way allows easily adding custom styles and overriding the existing ones.

dilirity commented on projects/js-packages/components/components/boost-score-graph/annotations-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+		annotations.forEach( annotation => {
+			const lineEl = document.createElement( 'div' );
+			lineEl.classList.add( 'jb-graph-annotations__line' );
+			lineEl.addEventListener( 'mouseenter', () => {
+				annotationEl.innerHTML = annotation.text;
+				annotationEl.style.display = 'block';
+				annotationEl.style.left = u.valToPos( annotation.timestamp / 1000, 'x' ) + 'px';
+			} );
+			lineEl.addEventListener( 'mouseleave', () => {
+				annotationEl.style.display = 'none';
+			} );
```
My native JS is rusty and I was wondering if it's possible to use one event listener for all.

When looping, add the annotation info as a data property so you can use it on mouse enter for the left position. Then add just one event listener to the parent element.

Just curious, could it be done with one event listener on the parent, instead of for every child? I might be mixing my jQuery days with the way the native JS works 😅

haqadn replied:
I think yes, you might be mixing it with jQuery. Each element will need it's own even listeners.
This is a bit different for the tooltips though. We always show the tooltip for nearest dataset, unless we are not hovering over the graph or highlighting one of these annotations.

dilirity reviewed (dismissed):
Great work! Happy new year to you too :D

There's just one small note about the JS/PHP schema, but other than that and the changelogger being unhappy, the rest is curiosities I had while reviewing.

thingalon commented:
This is a cool idea, and I am all for it but there are a lot of tasks in the TODO column for the React migration. Can we please pause this until the react migration is in hand?

pyronaur commented on projects/plugins/boost/wp-js-data-sync.php:
```diff
@@ -260,8 +260,16 @@ function jetpack_boost_initialize_datasync() {
 			),
-			'startDate' => Schema::as_number(),
-			'endDate'   => Schema::as_number(),
+			'annotations' => Schema::as_array(
+				Schema::as_assoc_array(
+					array(
+						'timestamp' => Schema::as_number(),
+						'text'      => Schema::as_string(),
+					)
+				)
+			)->nullable(),
+			'startDate'   => Schema::as_number(),
+			'endDate'     => Schema::as_number(),
```
Just making sure - is changing the schema with an already released feature not going to affect already existing data?

haqadn replied:
The data is coming from WPCOM, not being stored in the plugin. So, there shouldn't be any problem like that.

pyronaur replied:
And it's not cached in a transient either?

haqadn committed:
Merge branch 'trunk' into add/annotations

haqadn committed:
Make annotation not optional in boost context

It is still optional in the graph component in
RNA. But, boost will always pass an array
even if it's empty for type consistency.

haqadn committed:
Fixup project version

pyronaur commented on projects/js-packages/components/components/boost-score-graph/annotations-plugin.ts:
```diff
@@ -0,0 +1,70 @@
+		annotations.forEach( annotation => {
+			const lineEl = document.createElement( 'div' );
+			lineEl.classList.add( 'jb-graph-annotations__line' );
+			lineEl.addEventListener( 'mouseenter', () => {
+				annotationEl.innerHTML = annotation.text;
+				annotationEl.style.display = 'block';
+				annotationEl.style.left = u.valToPos( annotation.timestamp / 1000, 'x' ) + 'px';
+			} );
+			lineEl.addEventListener( 'mouseleave', () => {
+				annotationEl.style.display = 'none';
+			} );
```
Side note - make sure the event listeners don't leak, especially when changing routes without reloading the page. You might need a `removeEventListener` too

haqadn replied:
AFAIK, when we remove nodes from DOM, the even listeners also get removed. And, navigating between routes does remove and recreate the nodes, so we should always have new nodes with fresh set of event listeners. Am I missing something?

dilirity replied:
> If the event listener is not removed when the component unmounts, the event listener, its callback function, and your component will all stay in memory, causing a memory leak.

If you're adding one via the default API (addEventListener), you should always remove it.

Source: https://javascript.plainenglish.io/fixing-memory-leaks-in-react-e76e5d21632e

haqadn replied:
In this case, it is not a React component. The plugin is just another uPlot plugin that doesn't care if it's inside React or vanila JS. The destruction and re-creation of the graph are handled by uPlot's React wrapper, which handles destroying the [chart and thus nodes](https://github.com/skalinichev/uplot-wrappers/blob/master/react/uplot-react.tsx#L39). The chart is destroyed during unmounting.

haqadn committed:
Merge branch 'trunk' into add/annotations

thingalon reviewed (dismissed):
Seems to work, and I am happy with the implementation. I see the comments about leaving event listeners on DOM nodes -- absolutely good advice for React components, but these aren't React components.

haqadn committed:
Fixup project version
